src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TextDifferencing/TextDiffer.cs

```diff
@@ -18,38 +18,53 @@ internal abstract partial class TextDiffer
     protected abstract int OldSourceLength { get; }
     protected abstract int NewSourceLength { get; }
 
+    private int _oldSourceOffset;
+    private int _newSourceOffset;
+
     protected abstract bool SourceEqual(int oldSourceIndex, int newSourceIndex);
 
     protected List<DiffEdit> ComputeDiff()
     {
         var edits = new List<DiffEdit>(capacity: 4);
         var builder = new DiffEditBuilder(edits);
 
+        var lowA = 0;
+        var highA = OldSourceLength;
+        var lowB = 0;
+        var highB = NewSourceLength;
+
+        // Determine the extent of the changes in both texts, as this will allow us
+        // to limit the amount of memory needed in the vf/vr arrays. By doing this though,
+        // we will need to adjust SourceEqual requests to use an appropriate offset.
+        FindChangeExtent(ref lowA, ref highA, ref lowB, ref highB);
+
+        _oldSourceOffset = lowA;
+        _newSourceOffset = lowB;
+
+        var oldSourceLength = highA - lowA;
+        var newSourceLength = highB - lowB;
+
         // Initialize the vectors to use for forward and reverse searches.
-        var max = NewSourceLength + OldSourceLength;
+        var max = newSourceLength + oldSourceLength;
         using var vf = new IntArray((2 * max) + 1);
         using var vr = new IntArray((2 * max) + 1);
 
-        ComputeDiffRecursive(builder, 0, OldSourceLength, 0, NewSourceLength, vf, vr);
+        ComputeDiffRecursive(builder, 0, oldSourceLength, 0, newSourceLength, vf, vr);
+
+        // Update the resultant edits with the appropriate offsets
+        for (var i = 0; i < edits.Count; i++)
+        {
+            var edit = edits[i];
+
+            edits[i] = new DiffEdit(edit.Kind, _oldSourceOffset + edit.Position, _newSourceOffset + edit.NewTextPosition, edit.Length);
```

Comment:
Since there was effort to make the `DiffEdit` constructor private and provide factory methods, I'd prefer an `Offset` method that takes adjustments to `DiffEdit.Position` and `DiffEdit.NewTextPosition` rather than exposing the constructor and passing half of the same data back in.

Reply:
https://github.com/dotnet/razor/pull/11987